crates/http/src/http.rs

```diff
@@ -8,88 +9,44 @@ pub use isahc::{
     http::{Method, StatusCode, Uri},
     AsyncBody, Error, HttpClient as IsahcHttpClient, Request, Response,
 };
+use parking_lot::Mutex;
+use proxy::Proxy;
 #[cfg(feature = "test-support")]
 use std::fmt;
-use std::{
-    sync::{Arc, Mutex},
-    time::Duration,
-};
+use std::{sync::Arc, time::Duration};
 pub use url::Url;
 
-fn get_proxy(proxy: Option<String>) -> Option<isahc::http::Uri> {
-    macro_rules! try_env {
-        ($($env:literal),+) => {
-            $(
-                if let Ok(env) = std::env::var($env) {
-                    return env.parse::<isahc::http::Uri>().ok();
-                }
-            )+
-        };
-    }
-
-    proxy
-        .and_then(|input| {
-            input
-                .parse::<isahc::http::Uri>()
-                .inspect_err(|e| log::error!("Error parsing proxy settings: {}", e))
-                .ok()
-        })
-        .or_else(|| {
-            try_env!(
-                "ALL_PROXY",
-                "all_proxy",
-                "HTTPS_PROXY",
-                "https_proxy",
-                "HTTP_PROXY",
-                "http_proxy"
-            );
-            None
-        })
+/// Client Builder receives proxy settings and return an http client.
+pub trait ClientBuilder {
+    fn build(&self, proxy: Option<Uri>) -> Arc<dyn HttpClient>;
 }
 
-/// An [`HttpClient`] that has a base URL.
+/// An [`HttpClient`] that has a base URL and proxy settings.
 pub struct HttpClientWithUrl {
     base_url: Mutex<String>,
-    client: Arc<dyn HttpClient>,
-    proxy: Option<String>,
+    client_builder: Arc<dyn ClientBuilder + Send + Sync>,
+    proxy: Proxy,
 }
```

Comment:
Changing `proxy` field to some `Arc<Mutex<...>>` wrapped settings, and changing the `client` to a `client_builder` that consults the `proxy` field before every request.